File: cpp/tempexpr.cpp
#include <iostream>
#include "benjmark.h"
#include <vector>
#include <memory>
#include "common.h"

class ColumnVectorFromArray {
public:
  ColumnVectorFromArray(const std::vector<double>& src) : _n(src.size()), _data(src.data()) {}

  int rows() const {return _n;}
  int cols() const {return 1;}
  double get(int i, int j) const {
    return _data[i];
  }
private:
  const int _n;
  const double* _data;
};

template <typename T>
int numel(const T& src) {
  return src.rows()*src.cols();
}

int computeIndex(int rows, int i, int j) {
  return i + rows*j;
}

class Matrix {
public:
  Matrix(int rows, int cols) : 
    _rows(rows), _cols(cols), 
    _storage(std::make_shared<std::vector<double>>(rows*cols)) {
    _data = _storage->data();
  }

  int rows() const {
    return _rows;
  } 

  int cols() const {
    return _cols;
  }
  
  void setElement(int i, int j, double value) {
    _data[computeIndex(_rows, i, j)] = value;
  }

  double get(int i, int j) const {
    return _data[computeIndex(_rows, i, j)];
  }
private:
  int _rows;
  int _cols;
  double* _data;
  std::shared_ptr<std::vector<double>> _storage;
};

template <typename T>
Matrix realize(const T& mat) {
  int rows = mat.rows();
  int cols = mat.cols();
  Matrix dst(rows, cols);
  for (int i = 0; i < rows; i++) {
    for (int j = 0; j < cols; j++) {
      dst.setElement(i, j, mat.get(i, j));
    }
  }
  return dst;
}

template <typename T>
class Reshape {
public:
  Reshape(const T& src, int rows) : _src(src), _rows(rows), _cols(src.rows()/rows) {}

  int rows() const {return _rows;}
  int cols() const {return _cols;}
  
  double get(int i, int j) const {
    return _src.get(i + j*_rows, 0);
  }
private:
  int _rows;
  int _cols;
  T _src;
};

template <typename T>
Reshape<T> reshape(const T& src, int rows) {
  return Reshape<T>(src, rows);
}

template <typename T>
class Transpose {
public:
  Transpose(const T& src) : _src(src) {}  
  int rows() const {return _src.cols();}
  int cols() const {return _src.rows();}
  double get(int i, int j) const {
    return _src.get(j, i);
  }
private:
  T _src;
};

template <typename T>
Transpose<T> transpose(const T& x) {
  return Transpose<T>(x);
}

template <typename A, typename B>
class SubMat {
public:
  SubMat(const A& a, const B& b) : _a(a), _b(b) {}

  int rows() const {return _a.rows();}
  int cols() const {return _a.cols();}
  
  double get(int i, int j) const {
    return _a.get(i, j) - _b.get(i, j);
  }
private:
  A _a;
  B _b;
};

template <typename A, typename B>
SubMat<A, B> subMat(const A& a, const B& b) {
  return SubMat<A, B>(a, b);
}


template <typename A, typename B>
class MulMat {
public:
  MulMat(const A& a, const B& b) : _a(a), _b(b) {}

  int rows() const {return _a.rows();}
  int cols() const {return _b.cols();}
  
  double get(int i, int j) const {
    double sum = 0.0;
    int n = _a.cols();
    for (int k = 0; k < n; k++) {
      sum += _a.get(i, k)*_b.get(k, j);
    }
    return sum;
  }
private:
  A _a;
  B _b;
};

template <typename A, typename B>
MulMat<A, B> mulMat(const A& a, const B& b) {
  return MulMat<A, B>(a, b);
}

class Ones {
public:
  Ones(int rows, int cols) : _rows(rows), _cols(cols) {}
  int rows() const {return _rows;}
  int cols() const {return _cols;}
  double get(int i, int j) const {return 1.0;}
private:
  int _rows;
  int _cols;
};

template <typename T>
class ScaleMat {
public:
  ScaleMat(double s, const T& x) : _s(s), _x(x) {}

  int rows() const {return _x.rows();}
  int cols() const {return _x.cols();}

  double get(int i, int j) const {
    return _s*_x.get(i, j);
  }
private:
  double _s;
  T _x;
};

template <typename T>
ScaleMat<T> scaleMat(double s, const T& src) {
  return ScaleMat<T>(s, src);
}

Matrix covarianceMatrix(int dim, const std::vector<double>& data) {
  ColumnVectorFromArray V(data);
  auto X = reshape(V, dim);
  int N = data.size()/dim;
  auto mu = realize(scaleMat(1.0/N, mulMat(Ones(1, N), transpose(X))));
  auto muRepeated = transpose(mulMat(Ones(N, 1), mu));
  auto Xc = subMat(X, muRepeated);
  auto covariance = scaleMat(1.0/(N - 1), mulMat(Xc, transpose(Xc)));
  return realize(covariance);
}

struct Problem {
  int dim = 0;
  std::vector<double> data;
};

struct Setup {
  Problem input(const nlohmann::json& src) const {
    Problem dst;
    dst.dim = src["dim"];
    for (auto x: src["data"]) {
      dst.data.push_back(x);
    }
    return dst;
  }
  
  Matrix compute(const Problem& src) const {
    return covarianceMatrix(src.dim, src.data);
  }

  nlohmann::json output(const Matrix& src) const {
    nlohmann::json dst;
    for (int i = 0; i < src.rows(); i++) {
      nlohmann::json row;
      for (int j = 0; j < src.cols(); j++) {
        row.push_back(src.get(i, j));
      }
      dst.push_back(row);
    }
    return dst;
  }
  
};


int main(int argc, const char** argv) {
  Setup setup;
  CHECK(3 == argc);
  bj::perform(setup, argv[1], argv[2]);
  return 0;
}

File: srcjava/TempExpr.java
public class TempExpr {
    interface IExpr {
        int getRows();
        int getCols();
        double get(int i, int j);
    }

    class ColumnVectorFromArray implements IExpr {
        private double[] _data;
        
        ColumnVectorFromArray(double[] data) {
            _data = data;
        }

        public int getRows() {return _data.length;}
        public int getCols() {return 1;}

        public double get(int i, int j) {
            return _data[i];
        }
    }

    static int numel(IExpr expr) {
        return expr.getCols()*expr.getRows();
    }
    
    static int computeIndex(int rows, int i, int j) {
        return i + rows*j;
    }

    void dispMat(IExpr expr) {
        for (int i = 0; i < expr.getRows(); i++) {
            String s = "";
            for (int j = 0; j < expr.getCols(); j++) {
                s += expr.get(i, j) + " ";
            }
            System.out.println(s);
        }
    }

    class Matrix implements IExpr {
        double[] _data;
        int _rows;
        int _cols;
        
        Matrix(int rows, int cols) {
            _rows = rows;
            _cols = cols;
            _data = new double[rows*cols];
        }

        public int getRows() {return _rows;}
        public int getCols() {return _cols;}

        void set(int i, int j, double x) {
            _data[computeIndex(_rows, i, j)] = x;
        }

        public double get(int i, int j) {
            return _data[computeIndex(_rows, i, j)];
        }

        double[] getData() {return _data;}
    }

    Matrix realize(IExpr expr) {
        int rows = expr.getRows();
        int cols = expr.getCols();
        Matrix dst = new Matrix(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                dst.set(i, j, expr.get(i, j));
            }
        }
        return dst;
    }

    class Reshape implements IExpr {
        IExpr _src;
        int _rows;
        int _cols;

        Reshape(IExpr colvec, int rows) {
            _src = colvec;
            _rows = rows;
            _cols = colvec.getRows()/rows;
        }

        public int getCols() {return _cols;}
        public int getRows() {return _rows;}

        public double get(int i, int j) {
            return _src.get(i + _rows*j, 0);
        }
    }

    class Transpose implements IExpr {
        IExpr _src;

        Transpose(IExpr src) {
            _src = src;
        }

        public int getRows() {return _src.getCols();}
        public int getCols() {return _src.getRows();}

        public double get(int i, int j) {
            return _src.get(j, i);
        }
    }

    class SubMat implements IExpr {
        IExpr _a;
        IExpr _b;
        
        SubMat(IExpr a, IExpr b) {
            _a = a;
            _b = b;
        }

        public int getRows() {return _a.getRows();}
        public int getCols() {return _a.getCols();}
        

        public double get(int i, int j) {
            return _a.get(i, j) - _b.get(i, j);
        }
    }

    class MulMat implements IExpr {
        IExpr _a;
        IExpr _b;

        MulMat(IExpr a, IExpr b) {
            _a = a;
            _b = b;
        }

        public int getRows() {return _a.getRows();}
        public int getCols() {return _b.getCols();}

        public double get(int i, int j) {
            double sum = 0.0;
            int n = _a.getCols();
            for (int k = 0; k < n; k++) {
                sum += _a.get(i, k)*_b.get(k, j);
            }
            return sum;
        }
    }

    class Ones implements IExpr {
        int _rows;
        int _cols;
        
        Ones(int rows, int cols) {
            _rows = rows;
            _cols = cols;
        }

        public int getRows() {return _rows;}
        public int getCols() {return _cols;}
        
        public double get(int i, int j) {
            return 1.0;
        }
    }

    class ScaleMat implements IExpr {
        double _s;
        IExpr _x;

        ScaleMat(double s, IExpr x) {
            _s = s;
            _x = x;
        }

        public int getRows() {return _x.getRows();}
        public int getCols() {return _x.getCols();}
        
        public double get(int i, int j) {
            return _s*_x.get(i, j);
        }
    }

    Matrix covarianceMatrix(int dim, double[] data) {
        IExpr V = new ColumnVectorFromArray(data);
        IExpr X = new Reshape(V, dim);
        int N = data.length/dim;
        IExpr mu = realize(new ScaleMat(1.0/N, 
                new MulMat(
                    new Ones(1, N),
                    new Transpose(X))));

        IExpr muRepeated = new Transpose(
            new MulMat(
                new Ones(N, 1),
                mu));

        IExpr Xc = new SubMat(X, muRepeated);
        IExpr covariance = new ScaleMat(1.0/(N - 1), 
            new MulMat(Xc, new Transpose(Xc)));
        return realize(covariance);
    }

    public double[] run(int dim, double[] data) {
        return covarianceMatrix(dim, data).getData();
    }
    
}
